File: python/aethernet/bittorrent/merkle.py
# SPDX-License-Identifier: MIT
"""BitTorrent v2 (BEP-52) SHA-256 merkle hashing + v2 info-hash."""
from __future__ import annotations

import hashlib

BLOCK_SIZE = 16384


def merkle_root(data: bytes, block_size: int = BLOCK_SIZE) -> bytes:
    if block_size <= 0:
        raise ValueError("block size must be positive")
    leaves = [hashlib.sha256(data[i:i + block_size]).digest() for i in range(0, len(data), block_size)]
    if not leaves:
        return b"\x00" * 32
    return _root_of(leaves)


def _root_of(leaf_hashes: list[bytes]) -> bytes:
    level = list(leaf_hashes)
    width = 1
    while width < len(level):
        width <<= 1
    zero = b"\x00" * 32
    while len(level) < width:
        level.append(zero)
    while len(level) > 1:
        level = [hashlib.sha256(level[i] + level[i + 1]).digest() for i in range(0, len(level), 2)]
    return level[0]


def v2_info_hash(info_dict_bytes: bytes) -> bytes:
    return hashlib.sha256(info_dict_bytes).digest()


def v2_info_hash_truncated(info_dict_bytes: bytes) -> bytes:
    return v2_info_hash(info_dict_bytes)[:20]

File: kotlin/src/main/kotlin/aethernet/bittorrent/Merkle.kt
// SPDX-License-Identifier: MIT

package aethernet.bittorrent

import java.security.MessageDigest

/** The BitTorrent v2 leaf-block size (BEP-52). */
const val MERKLE_BLOCK_SIZE = 16384

private fun sha256(data: ByteArray): ByteArray =
    MessageDigest.getInstance("SHA-256").digest(data)

/**
 * Computes the SHA-256 merkle root of [data] split into 16 KiB leaf blocks (BEP-52):
 * each block SHA-256'd, the leaf layer zero-padded (a zero hash is 32 zero bytes) to
 * the next power of two, internal nodes SHA-256(left||right). Byte-identical to the
 * reference `MerkleTree.ComputeRoot`. The Kotlin port of `go/bittorrent/merkle.go`.
 */
fun merkleRoot(data: ByteArray): ByteArray = merkleRootBlock(data, MERKLE_BLOCK_SIZE)

/** [merkleRoot] with an explicit block size. */
fun merkleRootBlock(data: ByteArray, blockSize: Int): ByteArray {
    require(blockSize > 0) { "block size must be positive" }
    val leaves = ArrayList<ByteArray>()
    var i = 0
    while (i < data.size) {
        val end = minOf(i + blockSize, data.size)
        leaves.add(sha256(data.copyOfRange(i, end)))
        i += blockSize
    }
    if (leaves.isEmpty()) return ByteArray(32) // empty content → zero root
    return merkleRootOf(leaves)
}

/** Combines leaf hashes into a root, zero-padding to the next power of two. */
private fun merkleRootOf(leafHashes: List<ByteArray>): ByteArray {
    if (leafHashes.isEmpty()) return ByteArray(32)
    var level = ArrayList<ByteArray>(leafHashes)

    var width = 1
    while (width < level.size) width = width shl 1
    val zero = ByteArray(32)
    while (level.size < width) level.add(zero)

    while (level.size > 1) {
        val next = ArrayList<ByteArray>(level.size / 2)
        var j = 0
        while (j < level.size) {
            val combined = ByteArray(64)
            System.arraycopy(level[j], 0, combined, 0, 32)
            System.arraycopy(level[j + 1], 0, combined, 32, 32)
            next.add(sha256(combined))
            j += 2
        }
        level = next
    }
    return level[0]
}

/** The full 32-byte BitTorrent v2 info-hash: SHA-256 of the bencoded info dict. */
fun bitTorrentV2InfoHash(infoDictBytes: ByteArray): ByteArray = sha256(infoDictBytes)

/** The BitTorrent v2 info-hash truncated to 20 bytes. */
fun bitTorrentV2InfoHashTruncated(infoDictBytes: ByteArray): ByteArray =
    bitTorrentV2InfoHash(infoDictBytes).copyOfRange(0, 20)
